keras_nlp/utils/text_generation.py

```diff
@@ -0,0 +1,114 @@
+):
+    """
+    Text generation utility based on greedy search.
+
+    Greedy search always appends the token having the largest probability to
+    existing sequence.
+
+    Args:
+        token_probability_fn: a callable, which takes in input_sequence
+            and output the probability distribution of the next token.
+        input_ids: a list, the initial tokens to append generated tokens.
+        max_length: int. The max length of generated text.
+        end_token_id: int, defaults to None. The token marking the end of the
+            sequence, once encountered the generation is finished for the exact
+            sequence. If None, every sequence is generated up to `max_length`.
+
+    Returns:
+        A 1D int Tensor, or 2D int RaggedTensor representing the generated
+        sequences.
+
+    Examples:
+    ```python
+    VOCAB_SIZE = 10
+    FEATURE_SIZE = 16
+
+    # Create a dummy model to predict the next token.
+    model = tf.keras.Sequential([
+        tf.keras.Input(shape=[None]),
+        tf.keras.layers.Embedding(
+            input_dim=VOCAB_SIZE,
+            output_dim=FEATURE_SIZE,),
+        tf.keras.layers.Dense(VOCAB_SIZE, activation="softmax"),
+    ])
+
+    # Define a function that outputs the next token's probability given the
+    # input sequence.
+    def token_probability_fn(inputs):
+        return model(inputs)[:, -1, :]
+
+    inputs = tf.random.uniform(shape=[5, 5], maxval=VOCAB_SIZE, dtype=tf.int64)
+
+    # Print the generated sequence (token ids).
+    generate_text_greedy(
+        token_probability_fn,
+        inputs,
+        max_length=10,
+        end_token=0,)
+    ```
+
+    """
+    if 0 in input_ids.shape:
+        raise ValueError(
+            "input_ids must not be empty, but received empty input_ids."
+        )
+    input_is_1d = input_ids.shape.rank == 1
+    if input_is_1d:
+        input_ids = input_ids[tf.newaxis, :]
+
+    # Store if the end token has been generated for each sequence.
+    end_token_received = None
+    if end_token_id is not None:
+        end_token_received = tf.cast(
+            (input_ids[:, -1] == end_token_id), dtype=input_ids.dtype
+        )
+
+    def get_subsequent_tokens(input_ids, end_token_received):
+        if input_ids.shape[1] >= max_length:
+            # If the input_ids has reached our desired length, exit recursion.
+            return input_ids
+        pred = token_probability_fn(input_ids)
+        next_token = tf.cast(tf.argmax(pred, axis=-1), dtype=input_ids.dtype)
+        if end_token_id is not None:
+            # Replace the next token with `end_token_id` if end token has
+            # appeared in the sequenece.
+            next_token, end_token_received = _handle_end_token(
+                next_token, end_token_received, end_token_id
+            )
+        # Append the next token to current sequence.
+        input_ids = tf.concat([input_ids, next_token[:, tf.newaxis]], axis=-1)
```

Comment:
How about testing with XLA GPU at least if you cannot test on TPU :)?

Reply:
That's a good point. 

1) I am not sure how to add github GPU/TPU test. It also reminds that we might want to test distributed training (not for this utility but other modules) in the future. We will check more on it.
2) IIUC XLA on GPU should be turned on manually? It requires `@tf.function(jit_compile=True)`. For this specific utility, we probably don't need XLA testing?

Reply:
I am not sure if you realize that the tf.concat will yield dynamic shape and if you know xla requirements.

Reply:
Interesting - so tf.concat is something that cannot work on TPU and GPU/CPU (XLA)?

Our current plan is not to wrap this utility with `tf.function()`, but users can choose to wrap `next_token_fn` by `tf.function()`, as `next_token_fn` takes most of the computation time.

Reply:
the performance will be bad or you only target this as a demo util function? Is it the reason you use python while loop but not tf.while_loop for the decoding loop?

Reply:
Got it, yes, this util is mainly useful for a demo like colab guide, so performance is not the focus now. 

I am actually curious, how does model garden handle the token concatenation? Are you using a fixed size tensor, and change the value at each iteration? I am not sure how much performance diff it would have, since the bottleneck is mostly at model calling when I did benchmark on colab.

Reply:
We need to allocate buffer to the max sequence length to decode and use in-place update.
The padded_decode path is dedicated to XLA: https://github.com/tensorflow/models/blob/master/official/nlp/modeling/ops/beam_search.py#L109
BTW, I read fairseq code before and I guess they use something similar for GPU performance optimization as well.

Reply:
Thanks Hongkun! I will open an issue to track the refactoring,

Reply:
Yeah this isn't just not XLA compilable, it is not at all tf.function compilable right now.

We definitely should look at making it so, both for usability (use this inside a keras model) and performance for any sort of bulk inference like job.

For now, this has been more to get the API signature how we want it. I wouldn't say this should always be only a demo util function, that's just where we are at today.